job_templates/capsule_upgrade_-_ansible_default.erb

```diff
@@ -0,0 +1,58 @@
+<%#
+name: Capsule Upgrade Playbook
+snippet: false
+template_inputs:
+- name: target_version
+  required: true
+  input_type: user
+  advanced: false
+  value_type: plain
+  hidden_value: false
+- name: whitelist_options
+  required: false
+  input_type: user
+  advanced: false
+  value_type: plain
+  hidden_value: false
+model: JobTemplate
+job_category: Ansible Playbook
+description_format: "%{template_name}"
+provider_type: Ansible
+kind: job_template
+%>
+
+---
+- hosts: all
+  tasks:
+    - name: Gather the rpm package facts
+      package_facts:
+        manager: auto
+
+    - name: Fail if the target server is a Satellite server
+      fail:
+        msg: "This playbook cannot be executed on a Satellite server. Use only on a Capsule server."
+      when: "'satellite' in ansible_facts.packages"
+
+    - name: Install foreman-maintain if not present
+      package:
+        name: rubygem-foreman_maintain
+        state: present
+      when: "'rubygem-foreman_maintain' not in ansible_facts.packages"
+
+    - block:
+      - name: Upgrade Capsule server using foreman-maintain
+        command: foreman-maintain upgrade run --assumeyes --target-version=<%= input('target_version') %> --whitelist="<%= input('whitelist_options') %>"
+        register: result
+
+      - name: foreman-maintain upgrade return code is zero
+        debug:
+          msg: "Success! Capsule server upgrade completed. Current version of Capsule server server is {{ ansible_facts.packages['satellite-capsule'][0]['version'] }}."
+
+      rescue:
+        - name: Print foreman-maintain output
+          debug:
+            var: result
+
+        - name: foreman-maintain upgrade return code is non-zero
+          fail:
+            msg: "Failed! Capsule server upgrade failed. See /var/log/foreman-installer/capsule.log in the Capsule server for more information"
```

Comment:
could we actually `grep '\[E' /var/log/foreman-installer/capsule.log | head -n1` here to display on the WebUI the first installer issue that occurred?

does a non-zero return code always imply the failure is in the installer itself, or does foreman-maintain sometimes fail for other reasons?

Reply:
> could we actually do (some ansible equivalent of) `grep '\[E' /var/log/foreman-installer/capsule.log | head -n1` here to display on the WebUI the first installer issue that occurred?

Yes, I think this would help troubleshoot faster. I added the following now:
```console
        - name: Grep top 10 Error messages from /var/log/foreman-installer/capsule.log
          shell: grep '^\[ERROR' /var/log/foreman-installer/capsule.log | head -n10
          register: output_grep

        - name: Print grepped Error messages
          debug:
            var: output_grep.stdout_lines
```

> does a non-zero return code always imply the failure is in the installer itself, or does foreman-maintain sometimes fail for other reasons?

foreman-maintain can fail for pre or post checks too.

Reply:
same output of playbook with fake errors in capsule.log:
```console
TASK [Grep top 10 Error messages from /var/log/foreman-installer/capsule.log] ***
changed: [capsule-6-7-qa-rhel7.apex.example.com]

TASK [Print grepped Error messages] ********************************************
ok: [capsule-6-7-qa-rhel7.apex.example.com] => {
    "output_grep.stdout_lines": [
        "[ERROR 2015-01-05 14:00:29 main]  Nothing to do", 
        "[ERROR 2015-01-05 14:00:29 main]  Nothing to do", 
        "[ERROR 2015-01-05 14:00:29 main]  Nothing to do", 
        "[ERROR 2015-01-05 14:00:29 main]  Nothing to do", 
        "[ERROR 2015-01-05 14:00:29 main]  Nothing to do", 
        "[ERROR 2015-01-05 14:00:29 main]  Nothing to do"
    ]
}

Reply:
this is awesome @sthirugn , thanks!